flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java

```diff
@@ -540,14 +540,26 @@ protected int getCurrentOuterSnapshotVersion() {
 
         @Override
         protected void readOuterSnapshot(
-                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader) {
-            readVersion = readOuterSnapshotVersion;
+                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader)
+                throws IOException {
+            if (readOuterSnapshotVersion <= LAST_VERSION_WITHOUT_ROW_KIND) {
+                supportsRowKind = false;
+            } else if (readOuterSnapshotVersion == LAST_VERSION_WITHOUT_ROW_KIND + 1) {
+                supportsRowKind = true;
+            } else {
+                supportsRowKind = in.readBoolean();
+            }
+        }
+
+        @Override
+        protected void writeOuterSnapshot(DataOutputView out) throws IOException {
+            out.writeBoolean(supportsRowKind);
         }
 
         @Override
         protected OuterSchemaCompatibility resolveOuterSchemaCompatibility(
                 RowSerializer newSerializer) {
-            if (readVersion <= LAST_VERSION_WITHOUT_ROW_KIND) {
+            if (supportsRowKind == newSerializer.legacyModeEnabled) {
```

Comment:
I think we should better to add description to describe the relationship between `supportsRowKind` with `legacyModeEnabled`.